rust/noosphere-gateway/src/middleware.rs

```diff
@@ -0,0 +1,244 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use axum::{
+    extract::FromRequestParts,
+    headers::{authorization::Bearer, Authorization, Host},
+    http::{request::Parts, StatusCode},
+    TypedHeader,
+};
+use noosphere_core::{
+    api::headers as noosphere_headers,
+    authority::{SphereAbility, SphereReference, SPHERE_SEMANTICS, SUPPORTED_KEYS},
+    context::{HasMutableSphereContext, COUNTERPART},
+    data::Did,
+};
+use noosphere_storage::{KeyValueStore, Storage};
+use ucan::{capability::CapabilityView, chain::ProofChain, crypto::did::DidParser};
+
+use crate::ContextManager;
+
+/// Represents the scope of a gateway requests' access to a sphere.
+/// Use this in handlers to access a sphere context without authorizing.
+/// Otherwise, use [GatewayAuthority] which contains authorization
+/// as well as [GatewaySphereContext] functionality.
+pub struct GatewaySphereContext<M, C, S> {
```

Comment:
The naming and implementation here needs some additional thought. Here are my initial reactions:

1. We already have a thing called `SphereContext` and this wrapper is asymmetrical with it
2. The wrapper implementation appears to forward calls to `HasSphereContext` and `HasMutableSphereContext` trait methods, but does not implement those traits itself, which makes me wonder...
3. ...should this wrapper even exist? Can it just be replaced with any `impl HasMutableSphereContext` where it is referenced elsewhere? Could it just be a helper for constructing an `Arc<Mutex<SphereContext<...>>` or similar that is taken from it and passed on to the handlers?

Reply:
This was cleaned up significantly in the latest push, implementing directly `HasSphereContext` and `HasMutableSphereContext`. We can't implement `FromRequestParts` for e.g. `C`, hence the newtype (although technically not a new type because there's PhantomData)